firebase-common/src/main/java/com/google/firebase/FirebaseApp.java

```diff
@@ -305,6 +325,28 @@ public static FirebaseApp initializeApp(
     return firebaseApp;
   }
 
+  /**
+   * Specify which services should access local emulators for this FirebaseApp instance.
+   *
+   * <p>For example, if the {@link EmulatorSettings} contain {@link
+   * com.google.firebase.emulators.EmulatedServiceSettings} for {@link
+   * com.google.firebase.emulators.FirebaseEmulators#FIRESTORE}, then calls to Cloud Firestore will
+   * communicate with the emulator rather than production.
+   *
+   * <p>TODO(samstern): Un-hide this once Firestore, Database, and Functions are implemented
+   *
+   * @param emulatorSettings the emulator settings for all services.
+   * @hide
+   */
+  @KeepForSdk
+  public void enableEmulators(@NonNull EmulatorSettings emulatorSettings) {
+    checkNotDeleted();
+    Preconditions.checkState(
```

Comment:
How will this interact with firestore's `terminate()`? afaicr one can terminate one instance of firestore at which point the call to `FirebaseFirestore#getInstance()` will create a new firestore instance, should we allow the following sequence of calls?

```java
firestore.terminate();
app.updateFirestoresEmulator();
Firebase.firestore.use();
```

Reply:
Thanks for calling this out, I resolved all the other comments since they were straightforward but I will have to look into this one more deeply.

Reply:
Ok actually I do *not* want to allow that sequence of calls.  One of the main reasons for this is to make sure that app developers can reason about their emulation settings at a single point before *any* Firebase service has done any communication.

If you call `terminate()` the settings should still apply.  If you want to nuke them you'll need to call `FirebaseApp#delete()` and then make a new app.

WDYT?

Reply:
That's a good point and sgtm. You may want to confirm this behavior with firestore folks though

Reply:
Will do during the Firestore PR.